File: DataStructure(python)/graph/graph.py
'''
Created on 2017. 11. 6.

@author: Yoon
'''


def bfs(g, start):
    queue = [start]
    visited = []
        
    while queue:
        current = queue.pop(0)
        visited.append(current)
        for neighbor in g[current]:
            if neighbor not in visited:
                queue.append(neighbor)
    return visited            




def dfs(g, start):
    stack = [start]
    visited = []
    
    while stack:
        current = stack.pop()
        visited.append(current)
        while g[current]:
            neighbor = g[current].pop()
            if neighbor not in visited:
                stack.append(neighbor)
    return visited             
                
g = {
      0 : [1, 2],
      1 : [0, 3],
      2 : [0, 4, 5],
      3 : [1],
      4 : [2, 6],
      5 : [2],
      6 : [4]
    }                
                 
# g = {
#       1 : [2, 3],
#       2 : [1, 4, 5],
#       3 : [1],
#       4 : [2],
#       5 : [2]
#     }

# g = {'A': ['B', 'C', 'E'],
#      'B': ['A', 'D', 'F'],
#      'C': ['A', 'G'],
#      'D': ['B'],
#      'E': ['A', 'F'],
#      'F': ['B', 'E'],
#      'G': ['G']}

# g = {'세종': ['서울', '강릉', '대구', '광주'],
#          '서울': ['평양', '인천', '세종'],
#          '강릉': ['독도', '세종'],
#          '광주': ['세종', '여수'],
#          '대구': ['세종', '울산'],
#          '평양': ['서울', ],
#          '인천': ['서울', ],
#          '독도': ['강릉', ],
#          '여수': ['광주', '부산'],
#          '울산': ['대구', '부산'],
#          '부산': ['여수', '울산'],
#          }



print(bfs(g, 0))
print('------------------------------------')
print(dfs(g, 0))

File: DataStructure(JS)/Graph.js
/*
    자료구조 그래프
    BFS, DFS
 */

function Graph(){
    this.vertices = {};
}

function GraphNode(value){
    this.value = value;
    this.edge = [];
    this.visited = false;
}

// 그래프의 정점을 추가하는 메소드
Graph.prototype.addVertex = function(value){
    this.vertices[value] = new GraphNode(value);
}

// 간선을 추가하는 메소드
Graph.prototype.addEdge = function(v, w){
    this.vertices[v].edge.push(w);
    this.vertices[w].edge.push(v);
}

// 그래프 출력시
Graph.prototype.printGraph = function() {
    for (let key in this.vertices) {
        console.log(key + " --> " + this.vertices[key].edge);
    }
}

// 검색을 위해 원상 복구
Graph.prototype.resetVisited = function() {
    for (let key in this.vertices) {
        this.vertices[key].visited = false;
    }
}


Graph.prototype.bfs = function(start) {

    const queue = [start];
    const vertices = this.vertices;

    vertices[start].visited = true;

    while (queue.length) {
        let current = queue.shift();
        console.log(current);
        for (neighbor of vertices[current].edge) {
            if (!vertices[neighbor].visited) {
                vertices[neighbor].visited = true;
                queue.push(neighbor)
            }
        }
    }
}


Graph.prototype.dfs = function(start) {

    const vertices = this.vertices;
    vertices[start].visited = true

    console.log(vertices[start].value);

    for(neighbor of vertices[start].edge){
        if(!vertices[neighbor].visited){
            this.dfs(neighbor)
        }
    }
}


const graph = new Graph();
const myVertices = ['A', 'B', 'C', 'D', 'E', 'F'];
for (var i = 0; i < myVertices.length; i++) {
    graph.addVertex(myVertices[i]);
}


graph.addEdge('A', 'B');
graph.addEdge('A', 'D');
graph.addEdge('A', 'E');
graph.addEdge('B', 'C');
graph.addEdge('D', 'E');
graph.addEdge('E', 'F');
graph.addEdge('E', 'C');
graph.addEdge('C', 'F');

graph.printGraph();

graph.bfs('A');
graph.resetVisited();
graph.dfs('A');
